c2rust-analyze/src/main.rs

```diff
@@ -278,6 +278,53 @@ fn update_pointer_info<'tcx>(acx: &mut AnalysisCtxt<'_, 'tcx>, mir: &Body<'tcx>)
     }
 }
 
+fn foreign_mentioned_tys(tcx: TyCtxt) -> HashSet<DefId> {
+    let mut foreign_mentioned_tys = HashSet::new();
+    for ty in tcx
+        .hir_crate_items(())
+        .foreign_items()
+        .map(|item| item.def_id.to_def_id())
+        .filter_map(|did| match tcx.def_kind(did) {
+            DefKind::Fn | DefKind::AssocFn => Some(tcx.mk_fn_ptr(tcx.fn_sig(did))),
+            DefKind::Static(_) => Some(tcx.type_of(did)),
+            _ => None,
+        })
+    {
+        walk_adts(tcx, ty, &mut |did| foreign_mentioned_tys.insert(did));
+    }
+    foreign_mentioned_tys
+}
+
+/// Walks the type `ty` and applies a function `f` to it if it's an ADT
+/// `f` gets applied recursively to `ty`'s generic types and fields (if applicable)
+/// If `f` returns false, recursion terminates cases where the ADT is recursive,
+/// otherwise the function will naturally terminate given that the generic types
+/// of a type are finite in length.
+/// We only look for ADTs rather than other FFI-crossing types because ADTs
+/// are the only nominal ones, which are the ones that we may rewrite.
+fn walk_adts<'tcx, F>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, f: &mut F)
+where
+    F: FnMut(DefId) -> bool,
+{
+    // the first type encountered in the walk is `ty`, and the
+    // relevant handling for that is below, so we can skip it here
+    for arg in ty.walk().skip(1) {
+        if let GenericArgKind::Type(ty) = arg.unpack() {
+            walk_adts(tcx, ty, f);
+        }
+    }
```

Comment:
I think this can be exponential in the size of the type (which is usually small, but still).  Suppose you call `walk_adts(*mut *mut T)`: this loop will call `walk_adts(*mut T)` and `walk_adts(T)`.  Then the call to `walk_adts(*mut T)` will call `walk_adts(T)` a second time.  If `T = *mut *mut U`, then we call `walk_adts(U)` at least four times, and so on.

One easy fix for this would be to do only the `if let TyKind::Adt ...` work instead of recursing here.  For the types we visit in `walk()`, we don't want to walk them again (since the outer walk will already cover their children), we only want to do the `TyKind::Adt` check.

Reply:
Is 78fce6ab what you had in mind?

Reply:
Yes, except the `return` should be changed to `continue`

Reply:
Done